controllers/shooter.py

```diff
@@ -1,91 +1,91 @@
-from enum import Enum
-
-from magicbot import StateMachine, default_state, state
+from magicbot import StateMachine, default_state, state, timed_state, will_reset_to
+from wpimath.geometry import Pose2d
 
+from components.chassis import Chassis
 from components.intake import Intake
 from components.shooter import Shooter
 from components.tilt import Tilt
 from components.turret import Turret
-
-
-class GoalHeight(Enum):
-    HIGH = 2
-    MID = 1
-    LOW = 0
-
+from utilities.ballistics import GoalHeight, calculate_ballistics
 
 # Setpoints for intaking state
 INTAKE_AZIMUTH: float
 INTAKE_TILT: float
 
 
 class ShooterController(StateMachine):
+    chassis_component: Chassis
     intake_component: Intake
     shooter_component: Shooter
     tilt_component: Tilt
     turret_component: Turret
 
+    try_shoot = will_reset_to(False)
+
     def __init__(self) -> None:
-        # set default preference for goal height
-        pass
+        self.goal_height_preference = GoalHeight.HIGH  # default preference
 
     @state(first=True, must_finish=True)
     def preparing_intake(self) -> None:
-        # send pitch and turret to angle
+        self.turret_component.set_angle(0.0)
+        self.tilt_component.goto_intaking()
 
-        # if at pitch and turret angle
-        # next state intaking
-        pass
+        if self.turret_component.at_angle() and self.tilt_component.at_angle():
+            self.next_state("intaking")
 
     @state(must_finish=True)
     def intaking(self) -> None:
-        # shooter intake
-        # intake deploy
+        self.intake_component.deploy()
+        self.shooter_component.load()
 
-        # if shooter is loaded
-        # retract intake
-        # next state tracking
-        pass
+        if self.shooter_component.is_loaded():
+            self.intake_component.retract()
+            self.next_state("tracking")
 
     @default_state
     def tracking(self) -> None:
-        # if turret not indexed
-        # return
-
-        # update flywheel speed
-        # update turret angle
-        # update tilt angle
+        self.update_component_setpoints()
 
-        pass
+        if (
+            self.try_shoot
+            and self.shooter_component.is_ready()
+            and self.turret_component.at_angle()
+            and self.tilt_component.at_angle()
+        ):
+            self.next_state("shooting")
 
-    @state(must_finish=True)
+    @timed_state(must_finish=True, duration=1.0)
     def shooting(self) -> None:
-        # update flywheel speed
-        # update turret angle
-        # update tilt angle
-
-        # if tilt ready and turret ready and flywheels at speed
-        # shoot
-        # next state tracking
-        pass
+        self.update_component_setpoints()
+        self.shooter_component.shoot()
+
+    def update_component_setpoints(self) -> None:
+        bs = calculate_ballistics(
+            self.chassis_component.get_pose(), Pose2d(), self.goal_height_preference
```

Comment:
nit: maybe create a stub method for `calculate_target_pose` or whatever we want to call this second argument?

Reply:
Done